### PR TITLE
[CORDA-3628] - Implement sendAll API

### PR DESCRIPTION
**Pull request by dimosr**
Implementation of `sendAll` API to allow sending messages to multiple destinations with a single checkpoint.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 1]
```diff
@@ -317,6 +318,53 @@ abstract class FlowLogic<out T> {
...
+     * is offline then message delivery will be retried until the corresponding node comes back or until the message is older than the
+     * network's event horizon time.
+     *
+     * @param payload the payload to send.
+     * @param sessions the sessions to send the provided payload to.
+     * @param maySkipCheckpoint whether checkpointing should be skipped.
+     */
+    @Suspendable
+    @JvmOverloads
+    fun sendAll(payload: Any, sessions: List<FlowSession>, maySkipCheckpoint: Boolean = false) {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: Does `sessions` need to be a `Set` ?  What happens if you put the same `FlowSession` in more than once?

File: docs/source/api-flows.rst [hunk 2]
```diff
@@ -264,14 +264,18 @@ In order to create a communication session between your initiator flow and the r
 * ``sendAndReceive(receiveType: Class<R>, payload: Any): R``
     * Sends the ``payload`` object and receives an object of type ``receiveType`` back
 
-In addition ``FlowLogic`` provides functions that batch receives:
+In addition ``FlowLogic`` provides functions that can receive messages from multiple sessions and send messages to multiple parties:
```
Comment (rick-r3) [inline docs/source/api-flows.rst]: Stick with sessions or parties, rather than mixing them.  I think sessions is more accurate (could be 2 sessions with same party etc).

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt (hunk 1 above)
Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: That does raise the question of have we tested `receiveAll` with the same session in it more than once, since that seems to use a list?

File: docs/source/api-flows.rst [hunk 3]
```diff
@@ -264,14 +264,18 @@ In order to create a communication session between your initiator flow and the r
...
 
 * ``receiveAllMap(sessions: Map<FlowSession, Class<out Any>>): Map<FlowSession, UntrustworthyData<Any>>``
-  Receives from all ``FlowSession`` objects specified in the passed in map. The received types may differ.
+    * Receives from all ``FlowSession`` objects specified in the passed in map. The received types may differ.
 * ``receiveAll(receiveType: Class<R>, sessions: List<FlowSession>): List<UntrustworthyData<R>>``
-  Receives from all ``FlowSession`` objects specified in the passed in list. The received types must be the same.
+    * Receives from all ``FlowSession`` objects specified in the passed in list. The received types must be the same.
+* ``sendAll(payload: Any, sessions: List<FlowSession>)``
+    * Sends the ``payload`` object to all the provided ``sessions``.
```
Comment (rick-r3) [inline docs/source/api-flows.rst]: to all the provided ``FlowSession``s (for consistency with receiveAll description).

File: docs/source/api-flows.rst [hunk 4]
```diff
@@ -264,14 +264,18 @@ In order to create a communication session between your initiator flow and the r
...
+* ``sendAll(payloadsPerSession: Map<FlowSession, Any>)``
+    * Sends one payload to every session, as specified by the provided ``payloadsPerSession``.
```
Comment (rick-r3) [inline docs/source/api-flows.rst]: Sends a potentially different payload to each `FlowSession`...

File: docs/source/api-flows.rst [hunk 5]
```diff
@@ -264,14 +264,18 @@ In order to create a communication session between your initiator flow and the r
...
 
-The batched functions are implemented more efficiently by the flow framework.
+.. note:: It's more efficient to call ``sendAndReceive``, instead of calling ``send`` and then ``receive``. It's also more efficient to call ``sendAll``, instead of multiple ``send``\s.
```
Comment (rick-r3) [inline docs/source/api-flows.rst]: Likewise receiveAll ?

File: node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt [hunk 6]
```diff
@@ -53,6 +53,11 @@ class MessagingExecutor(
         producer.send(SimpleString(mqAddress), artemisMessage)
     }
 
+    @Synchronized
+    fun send(messages: Map<MessageRecipients, Message>) {
+        messages.forEach { recipients, message -> send(message, recipients) }
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt:58]: Is this so we could later optimise within `MessagingExecutor` ?

File: node/src/main/kotlin/net/corda/node/services/statemachine/transitions/StartedFlowTransition.kt [hunk 7]
```diff
@@ -246,46 +247,49 @@ class StartedFlowTransition(
...
-                                actions.add(Action.SendExisting(existingSessionState.peerParty, existingMessage, SenderDeduplicationId(deduplicationId, startingState.senderUUID)))
-                                Unit
-                            }
-                            InitiatedSessionState.Ended -> {
-                                return freshErrorTransition(IllegalStateException("Tried to send to ended session $sourceSessionId"))
-                            }
-                        }
-                    }
-                }
+        for ((sourceSessionId, _) in sourceSessionIdToMessage) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/transitions/StartedFlowTransition.kt:250]: Note to self:  Okay, I reviewed this whole method and it has the same logic as before, but it packages up the individual actions into a "batch" that it then incorporates into a `SendMultiple` action.  I need to understand what happens then, but the logic here is otherwise unchanged.

File: node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt [hunk 8]
```diff
@@ -191,6 +192,13 @@ class ActionExecutorImpl(
         flowMessaging.sendSessionMessage(action.peerParty, action.message, action.deduplicationId)
     }
 
+    @Suspendable
+    private fun executeSendMultiple(action: Action.SendMultiple) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/ActionExecutorImpl.kt:196]: Right, so the multiple send allows you to call the batched send in the messaging layer. Got it.

**Review by rick-r3 — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt (hunk 1 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: Good spot. 
- This delegates to the method that uses a map, so sessions will be deduplicated naturally. But, I agree that it would be better to have a set here to be clear to the user. I'll switch to that.
- The `receiveAll` explicitly enforces no duplicates and will fail with an `IllegalArgumentException`, if there is any. It seems there's no test that verifies exactly that, I can add one if you think it's worth it.

File: docs/source/api-flows.rst (hunk 2 above)
Comment (dimosr) [reply docs/source/api-flows.rst]: yeah, agree will switch to sessions.

File: node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt (hunk 6 above)
Comment (dimosr) [reply node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt:58]: Yeah, based on your comments I understand you figured out later. But, I tried to keep it as a batch down to the lower level mostly to reduce code divergence between OS/ENT and make it easy to optimise batch at this level.

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt (hunk 1 above)
Comment (mnesbit) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: In the world with Accounts I think we do have to identify what is the expected behaviour of self-sessions and duplicate sessions. Especially as we know that the customer is doing these things in their flows

Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I think we should add a unit test for the `receiveAll` with duplicates.  Sounds like an easy one to lock in a test for that behaviour.

Self-sessions work just fine.  They behave just as if you talk to a different node.  Always have done in the new state machine.  It's logic in the built in flows that has the potential to confuse.  Duplicate sessions are only relevant here and are not supported simply because the code does not support them (hence the proposed switch to `Set` in the API).  We could make it support them but no intention of creeping the  scope.

File: node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt (hunk 6 above)
Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/messaging/MessagingExecutor.kt:58]: Yeah, I worked it out once I reviewed all the code change.  I agree this is the way to go.

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt (hunk 1 above)
Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: I added a test for `receiveAll` with duplicates. I also could not find a test for self-session messaging, so I added one for this too.

Btw, even if we had the resources/time, I am not sure supporting duplicates at this level would be a good decision design-wise - the semantics could end up being quite confusing.

Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: Are you still intending to switch this `sendAll` to take a `Set` in the method signature ?

Comment (mnesbit) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: Just has to follow same rules as CollectSignaturesFlow I guess

Comment (dimosr) [reply core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: @rick-r3 of course, I forgot it (which worries me a bit tbh..)

**Review by rick-r3 — APPROVED** [review]
LGTM
